1 3 feature cart

ermlvgrsh opened the pull request:
Выбранная архитектура: MVVM
 Выбранная верстка: верстка кодом
 Эпик - Корзина
 ссылка на таблицу: https://github.com/users/igor-ignatov/projects/1/views/9?filterQuery=%5BCart%5D+
Эпик - Корзина
Выполнено
 UI задачи
1	Верстка экрана корзины с двумя стэйтами (1. корзина пуста; 2. в корзине есть продукция) 
Создание placeholder view
 Создание tableView с продукцией 
⁃ а) создание ячейки TableView которая включает в себя
 ⁃ название продукции 
⁃ вью с рейтингом
 ⁃ цена продукции 
⁃ кнопки удаления Создание кнопки сортировки
 Создание стэк вью с кнопкой об оплате
 Планируемое время 6 часов
 Фактическое время 14 часов
2	Реализовать кастомное отображение контекстного меню для удаления ячейки таблицы Создание отдельного вью с эффектом VisualEffectsView
 вывести imageView ячейки
	•	лэйбл с вопросом
	•	две кнопки - удаление и отмена 
Планируемое время 3 часа
Фактическое время 5 часов
3	Верстка экрана способа оплаты
 Создать и настроить collectionView
 Создать ячейки для коллекшнВью в которых содержится
 картинка, полное имя, сокращенное имя 
Планируемое время 6 часов
Фактическое время 6 часов
4  Реализовать сервис по получению позиции из каталога
	•	Реализовать CartManager для работы с сетью
	•	Реализовать модель корзины
	•	Реализовать вьюмодел для работы с вьюконтроллером 
Планируемое время 3 часа
Фактическое время 6 часов

5 Реализовать сортировку NFT
 Планируемое время 3 часа
Фактическое время 3 часа

Необходимо сделать
Верстка экрана результата оплаты с двумя стэйтами (1. Оплата прошла успешна; 2. Оплата не прошла) Планируемое время 3 часа
Реализовать переход на WebView для пользовательского соглашения
 Планируемое время 2 часа
 Бизнес логика
Реализовать сервис по обновлению позиций при изменении заказа
 Планируемое время 3 часа
Реализовать сервис по получения всех возможных валют + добавить возможность оплаты заказа
 Планируемое время 4 часа

ermlvgrsh committed:
implement cart view controller

ermlvgrsh committed:
implement custom views
implement view model + model

ermlvgrsh committed:
implement Purchase Controller
implement Constants
add AgreementViewController
add Network Request + CartManager

ermlvgrsh committed:
Create epic_cart_plan

Helvetii951 commented on FakeNFT/Scenes /Cart/CartPurchase/View/CartPurchaseViewController.swift:
```diff
@@ -0,0 +1,132 @@
+extension CartPurchaseViewController: UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return 8
+    }
```
Желательно не применять [магические числа](https://bureau.ru/soviet/20230406/). Такие числовые переменные лучше выностить в свойства или структуры с понятными названиями. Что-то вроде numberOfCurrencies

Helvetii951 commented on FakeNFT/Scenes /Cart/CartPurchase/View/CartPurchaseViewController.swift:
```diff
@@ -0,0 +1,132 @@
+            withReuseIdentifier: CartPurchaseCell.identifier,
+            for: indexPath) as? CartPurchaseCell else { fatalError() }
+        cell.configureCell()
+        return cell
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfailure(_:file:line:))

Helvetii951 commented on FakeNFT/Scenes /Cart/CartPurchase/View/WebViewController/AgreementViewController.swift:
```diff
@@ -0,0 +1,39 @@
+        let webView = WKWebView()
+        webView.translatesAutoresizingMaskIntoConstraints = false
+        return webView
+    }()
```
Здесь можно использовать [sfsafariviewcontroller](https://developer.apple.com/documentation/safariservices/sfsafariviewcontroller). В несколько строчек можно создать экземпляр, передать урл и отобразить веб-страницу. Плюс мы получим плюшки - почти все, что есть в Safari.

Helvetii951 commented on FakeNFT/Scenes /Cart/CartViewCell/CartTableViewCell.swift:
```diff
@@ -0,0 +1,121 @@
+    private let nftImageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.layer.masksToBounds = true
+        imageView.image = UIImage(named: "AppIcon")
```
Все же рекомендую использовать библиотеки генерации ресурсов типа [SwiftGen](https://github.com/SwiftGen/SwiftGen). Помогает решить очень много проблем с ресурсами.

Helvetii951 commented on FakeNFT/Scenes /Cart/CartViewController.swift:
```diff
@@ -1,3 +1,229 @@
+        cartTableView.dataSource = self
+    }
+    
+    private func bindViewModel() {
+        viewModel.$cartModels.bind { [weak self] _ in
+            self?.cartTableView.reloadData()
+        }
+        viewModel.$isTableViewHidden.bind { [weak self] isHidden in
+            self?.cartTableView.isHidden = isHidden
+        }
+        viewModel.$isPlaceholderHidden.bind { [weak self] isHidden in
+            self?.placeholderLabel.isHidden = isHidden
+        }
+        viewModel.bindCart()
+    }
```
Биндинги хорошо реализованы через обертки свойств👍

Рекомендую уже сейчас начинать знакомиться с реактивщиной. Концепция не очень простая, но супер сильная. Эти знания скоро будут необходимы в связке SwiftUI + Combine. Плюс такие знания станут преимуществом при трудоустройстве.

[Документация Apple по фреймворку Combine](https://developer.apple.com/documentation/combine)
[Неплохая вводная статья](https://habr.com/ru/companies/simbirsoft/articles/712332/)

Helvetii951 commented on FakeNFT/Scenes /Cart/CustomView/DeleteView.swift:
```diff
@@ -0,0 +1,141 @@
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    private let bluredBackgroundView: UIVisualEffectView = {
+        let effect = UIBlurEffect(style: .regular)
+        let view = UIVisualEffectView(effect: effect)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
```
👍

Helvetii951 commented on FakeNFT/Scenes /Cart/CartViewController.swift:
```diff
@@ -1,3 +1,229 @@
+
+final class CartViewController: UIViewController {
+    private var indexCell: Int?
+    let viewModel: CartViewModel
+    init(viewModel: CartViewModel) {
+        self.viewModel = viewModel
+        super.init(nibName: nil, bundle: nil)
+    }
```
Рекомендую объявить вью-модель не конкретной реализацией, а протоколом.

[Хорошая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/)

ermlvgrsh committed:
correct tableView
redo collectionView
add swiftLint rule
redo WebView into SFSafari and add Agreement
